libs/core/algorithms/include/hpx/parallel/datapar/adjacent_find.hpp

```diff
@@ -0,0 +1,130 @@
+//  Copyright (c) 2021 Srinivas Yadav
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/config.hpp>
+
+#if defined(HPX_HAVE_DATAPAR)
+#include <hpx/concepts/concepts.hpp>
+#include <hpx/execution/traits/is_execution_policy.hpp>
+#include <hpx/execution/traits/vector_pack_all_any_none.hpp>
+#include <hpx/execution/traits/vector_pack_find.hpp>
+#include <hpx/executors/datapar/execution_policy.hpp>
+#include <hpx/functional/tag_invoke.hpp>
+#include <hpx/parallel/algorithms/detail/adjacent_find.hpp>
+#include <hpx/parallel/datapar/iterator_helpers.hpp>
+#include <hpx/parallel/datapar/loop.hpp>
+#include <hpx/parallel/datapar/zip_iterator.hpp>
+#include <hpx/parallel/util/result_types.hpp>
+
+#include <cstddef>
+#include <type_traits>
+#include <utility>
+
+namespace hpx { namespace parallel { inline namespace v1 { namespace detail {
+
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename ExPolicy>
+    struct datapar_adjacent_find
+    {
+        template <typename InIter, typename Sent_, typename PredProj>
+        static inline InIter call(
+            InIter first, Sent_ last, PredProj&& pred_projected)
+        {
+            if (first == last)
+                return last;
+
+            InIter next = first;
+            ++next;
+
+            auto zip_iter = hpx::util::make_zip_iterator(first, next);
+            std::size_t count = std::distance(first, last);
```

Comment:
`std::distance` will not work if `!std::is_same_t<InIter, Sent_>`, please consider `parallel::detail::distance` instead.